### PR TITLE
[event] trigger events processing

### PR DESCRIPTION
**Pull request by longbowlu**
This PR mainly does 3 things:
1. implement `EventHandler` and add as a field of `AuthorityState`
2. bootstrap `Streamer` which handles events streaming and subscription managements
3. trigger events processing  

This PR include changes in https://github.com/MystenLabs/sui/pull/2324

### CODE REVIEW

File: crates/sui-core/src/event_handler.rs
Comment (velvia) [inline crates/sui-core/src/event_handler.rs:22]: We want the UTC timestamp, and 64 bits is sufficient, we don't need 128.  

Instead of the above I advise using `Utc::now().timestamp()` from `use chrono::prelude::*`
(chrono is already a dependency of multiple packages)

File: crates/sui-core/src/event_handler.rs
Comment (velvia) [inline crates/sui-core/src/event_handler.rs:60]: So, getting the move struct out of the event will also be needed by the path to write to event store also.  Could we do the event handler in a reactive / async stream style, then we can just have a chain of event processing steps in the stream?  The first one could be adding the move struct, or we could add it via populating a OnceCell based field in MoveEvent for the struct.  I could then use the output of this first step in the EventStore.

File: crates/sui-types/src/event.rs
Comment (velvia) [inline crates/sui-types/src/event.rs:40]: I don't think the move_event_json_contents thing should be pre-populated the first time.... there are better ways to do this as it might be expensive.  We can use OnceCell, so that it's lazy, OR better yet maybe we can get Move to give us the original MoveStruct without deserializing it.

**Review by velvia — CHANGES_REQUESTED** [review]
Just some thoughts, this doesn't quite look finished?  :)

File: crates/sui-core/src/event_handler.rs
Comment (longbowlu) [reply crates/sui-core/src/event_handler.rs:60]: I see, maybe I can add `MoveStruct` in `EventEnvolope` too if it's useful afterwards

regarding streaming, I didn't make it that way because the pre-processing part (mainly`extract move struct`) seems to be trivial. Do you just prefer the streaming style or do you see other advantages by streaming earlier on? (In case you missed it, I was thinking to interact with event store in line 73)

The event dissemination is designed in a streaming way, see line 81-86

**Review by longbowlu — COMMENTED** [review]
@velvia  it's indeed "WIP", the major missing part is `Streamer` which does filtering and deliver events. But I don't want to this PR to blocked by that because:
1. once this PR lands you can integrate with EventStore
2. `Streamer` depends on the subscription endpoints that Patrick is working on
3. smaller PR is easier to review

File: crates/sui-core/src/event_handler.rs
Comment (velvia) [reply crates/sui-core/src/event_handler.rs:60]: So basically the streaming style - well I guess part of it is preference, but part of it is a easy way to bulk ingest events and take advantage of processing multiple events.  Like, we don't want to just ingest one event at a time, but at minimum to ingest all the events for a single TX.  Better yet though is to ingest many events at the same time.

Also if you handle each type of event differently how can you be sure that the order is preserved in the stream?   Keeping things as a stream makes sure we still preserve ordering.

File: crates/sui-types/src/event.rs
Comment (velvia) [reply crates/sui-types/src/event.rs:40]: > Yeah I can include MoveStruct here. But we probably still need move_event_json_contents as it's useful for subscription filtering. I will take another look to confirm though.

Two things:
1. It's more efficient to filter using MoveStruct than JSON representation, in particular if you want to filter by things like ObjectID, Address, etc. because doing binary comparisons is much cheaper than the string (or hex representation) comparisons.
2. The primary fields we are trying to filter by, including the package and module ID, and event type, are not in the MoveStruct that is deserialized, but rather from the MoveLayout.  In fact those are fields that are in other event types too, not just the user MoveEvents.  

2 is the reason that I say, maybe the move struct / json stuff should be optional.

Also, move struct -> JSON is very cheap, the expensive part is deserializing raw bytes into a MoveStruct, I'm pretty sure.   Well, I guess we can measure it.  

I'm also suggesting to standardize on MoveStruct because from a storage perspective, it is better to store something closer to the original structure, for example preserving binary types and blobs.  The JSON representation is not as canonical, like there is for example no canonical way to represent binary data in JSON.  That said there are advantages to JSON as well,

File: crates/sui-core/src/event_handler.rs
Comment (longbowlu) [reply crates/sui-core/src/event_handler.rs:53]: actually async is needed because of sending events to streamer. See line 80

File: crates/sui-types/src/event.rs
Comment (longbowlu) [reply crates/sui-types/src/event.rs:40]: resolved offline.

FWIW, json is more filtering friendly because we can easily say "json_val['foo']['bar']". For MoveStruct we have to invent the wheels by ourselves, which I would punt unless there's a strong incentive to do so

File: crates/sui-types/src/event.rs
Comment (longbowlu) [reply crates/sui-types/src/event.rs:32]: resolved offline.

regarding `OnceCell`, I wanted to use it but it would be a bit ugly because the necessary `ModuleResolver` is not part of the struct. I will leave it as it is

File: crates/sui-core/src/authority.rs
Comment (velvia) [inline crates/sui-core/src/authority.rs:530]: I had another thought but we can address this in a separate PR.   I don't think we really need to have all the event processing completed as a part of the normal call to process a certificate, basically this adds latency to certificate processing, when clients don't really need event processing to be completed before the call returns.  So in the future we should move this out to be done in the side, like push to a stream or channel or something.

File: crates/sui-core/src/event_handler.rs
Comment (velvia) [inline crates/sui-core/src/event_handler.rs:22]: Thanks for the change.  BTW I've never seen i64 -> u64 fail so you could probably just do `ts_ms as u64`

File: crates/sui-core/src/streamer.rs
Comment (velvia) [inline crates/sui-core/src/streamer.rs:18]: This seems like a temporary/debug thing, to be replaced by real code. Let's put a TODO here.

File: crates/sui-core/src/event_handler.rs
Comment (velvia) [inline crates/sui-core/src/event_handler.rs:71]: I would be a bit worried about performance implications of always waiting to send after every single event, but maybe it's fairly quick.  We can measure and improve on it later.

File: crates/sui-core/src/event_handler.rs
Comment (velvia) [inline crates/sui-core/src/event_handler.rs:72]: This should be really rare right?    Sending to a queue should typically not fail.   I wonder if we need a separate error for this.

**Review by velvia — APPROVED** [review]
Thanks @longbowlu looks good, just left some minor comments.  We can iterate on this in subsequent PRs.

File: crates/sui-core/src/authority.rs
Comment (longbowlu) [reply crates/sui-core/src/authority.rs:530]: certainly

File: crates/sui-core/src/event_handler.rs
Comment (longbowlu) [reply crates/sui-core/src/event_handler.rs:22]: it can fail when it's negative numbers (hence "travelling in time machine")

File: crates/sui-core/src/streamer.rs
Comment (longbowlu) [reply crates/sui-core/src/streamer.rs:18]: oh yeah I'm going to work on this after this is merged

File: crates/sui-core/src/event_handler.rs
Comment (longbowlu) [reply crates/sui-core/src/event_handler.rs:72]: The only failing scenario is the streamer dies then the receiver handle is dropped. I can't find an existing error to represent such a case

**Pull request merged**
